DM-14289

timothydmorton opened the pull request:
This contains the implementations of `PostprocessAnalysis` and `PostprocessTask`, which enable general postprocessing computations on output catalogs, and the ability to define such with a yaml file.  The `QADataset` object is also rewritten, which enables plotting under this new paradigm.

timothydmorton committed:
Update functors to work with ParquetTable objects, and add tests

Before, the __call__ method of a `Functor` expected to take a `Catalog` as argument; this changes things so that it takes a `ParquetTable` object. One of the important things that change here is a functor now specifies a "filt" and "dataset" that it acts upon.  This can be specified upon declaration of a Functor, or not specified, in which case default values are used (defaults specified in functor definition).
This commit also includes changes to how `from_yaml` parses a yaml functor specifying a `CompositeFunctor`, so that it parses abitrary kwargs, and takes a 'flags' section without specifying what dataset to be taken from (in which case the flags are taken from the 'ref' dataset).
In addition, this makes dropping levels optional when calling the
`.toDataFrame` method of `MultilevelParquetTable`.  This is necessary
due to the new behavior of the functors, in particular CompositeFunctor.

timothydmorton committed:
Add MagErr, NanoMaggie, and Index functors

timothydmorton committed:
Rewrite QADataset to take just a simple dataframe

Before, initializing a `QADataset` required passing a catalog and a group of functors, and the QADataset object would manage the computation of the functor columns on that dataset.  This commit changes QADataset to simply receive an already-calculated dataframe, taking away all calculation functionality, and just managing the creation and wrapping of the holoviews `Dataset` object; e.g., by hiding away things like defining the kdims and vdims, getting rid of null values, etc.

This commit also keeps the old version of dataset.py as dataset_old.py.
This is a temporary convenience to keep all the old stuff visible in the
current repo until all the old functionality is replicated, at which
point it will be removed.

timothydmorton committed:
Implement PostprocessAnalysis and PostprocessTask objects; remove writeQATable.py

The `PostprocessAnalysis` object takes over the calculation-of-columns-from-catalog functionality that used to be within QADataset.  This is a generalized object that takes a `ParquetTable` object and a list/dict of functors (or a CompositeFunctor).  After initialization, accessing the `.df` attribute triggers computation.  `PostprocessTask` is a general purpose task that uses the PostprocessAnalysis object to compute a new output dataset given a YAML file that specifies a CompositeFunctor (`--functorFile` config parameter).  `MultibandPostprocessTask` computes the requested set of functors for each filter.  This completely replaces `WriteQATableTask` and the old `PostprocessCatalogTask`, which were defined in `writeQATable.py`, which is removed by this commit.

timothydmorton committed:
Replace IcrsCoord with SpherePoint

timothydmorton committed:
Correct typo bug: self.pf -> self._pf

timothydmorton committed:
Add some error checking to plots functions; remove lots of whitespace

timothydmorton committed:
Use rasterize to enable colorbar for sky plots.

The `holoviews.operations.datashader.datashade` operation does not allow for a colorbar; however, `rasterize` does.  This changes the "sky" part of the "scattersky" plots to use rasterize and thus enable a colorbar.  The downside to this is that some of the functionality of 'datashade' goes away, such as the ability to use `dynspread` to increase the size of "points" for clarity when zoomed in.

timothydmorton committed:
Add warning that auto-generated notebooks will be obselete until further notice.

timothydmorton committed:
Minor changes for backward compatibility of old-style interactive plots (e.g., with forced.parq, etc.)

This file will probably be removed shortly anyway, as `ParquetTable` + `PostprocessAnalysis` is slated to completely replace the functionality of `Catalog` + `QADataset`, though this process is not yet complete.

timothydmorton committed:
Add documentation to functors

timothydmorton committed:
fix path name to test YAML file

timothydmorton committed:
correct QAfunctors.yaml for new parsing

timothydmorton committed:
extend functor tests to multiple filters

timothydmorton committed:
documentation for postprocess.py

timothydmorton committed:
documentation for dataset.py

timothydmorton committed:
change run to runDataRef, etc.

timothydmorton committed:
add new demo notebook, semi-hide old ones

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-from .plots import filter_dset
+from lsst.qa.explorer.match import match_lists
+from lsst.qa.explorer.plots import filter_dset, FilterStream
 
```
expected 2 blank lines

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-    labeller : explorer.functors.Labeller
-        Functor to assign labels to sources.  Default is 
-        `explorer.functors.StarGalaxyLabeller()`.
+    * Defining the `hv.Dataset` object, which beyond defining the dimensions
```
Want to hear more about necessitating getting rid of infs and nans... How do you do that? Do you lose whole rows? when does this happen? etc

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-    oom : bool
-        Whether to store computed dataframe out of memory.  Future
-        to-be-implemented feature; not really used yet.
+    vdims : `str`, optional
```
How should this string be formatted? List OK?

timothydmorton replied:
This should actually be a list of strings.

timothydmorton replied:
(or `'all'`, that is)

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-            Dictionary or list of functors.  If list, then will be 
-            converted into dictionary keyed by `y0, y1, y2, ...`.
 
+    def __init__(self, df, vdims='all'):
```
expected 1 blank line

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
     @property
     def df(self):
-        """Dataframe containing results of computation
+        """Dataframe
```
More?

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+        """
+        if self._flags is None:
+            self._flags = [c for c in self.df.columns
+                            if self.df[c].dtype == np.dtype('bool')]
```
bad indent

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+            if (c in self._kdims or
+                c in self._idNames or
+                c in self.flags):
+                kdims.append(c)
```
This either needs reformatting or an `ignore E129` to be allowed into pipe_tasks. Oh but this one isn't going into pipe_tasks.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-            self._df_file = os.path.join(self._cachedir, next(tempfile._get_candidate_names()))
-        return self._df_file
+    def vdims(self):
+        _, vdims = self._getDims()
```
Is this safer than `return self._getDims()[1]`? It is clear as is.

timothydmorton replied:
Not that I know; I just like the style for some reason.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+    def _makeDataset(self):
+        kdims, vdims = self._getDims()
 
+        df = self.df.replace([np.inf, -np.inf], np.nan).dropna(how='any')
```
Oh yikes. This might surprise people.

timothydmorton replied:
This is to facilitate plotting with holoviews; at least when I was developing this, infs or nans would cause the plots to barf, so I get rid of them.  The `hv.Dataset` object created here (`self.ds`) is only the support vehicle for the plots, and the underlying `self.df` is unaffected by this.  Does this also answer your above question about infs/nans?

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+        self._ds = ds
 
-    def visit_points(self, vdim, visit, x_max, label,
+    def skyPoints(self, vdim, maxMag, label='star', magCol='psfMag',
```
This rename requires a whitespace change below. What are the options for `label`?

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-            Parameters controlled by the `filter_stream` parameter
-            of `visit_explore`.
-        """
+        selectDict = {magCol : (0,maxMag),
```
Docstring. Remove whitespace before `:`

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-            of `visit_explore`.
-        """
+        selectDict = {magCol : (0,maxMag),
+                       'label' : label}
```
whitespace

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+        pts = hv.Points(ds, kdims=['ra', 'dec'], vdims=self.vdims + [magCol] + self.idNames)
+        return pts.options(color_index=vdim)
+
+    def skyDmap(self, vdim, magRange=(np.arange(16,24.1, 0.2)), magCol='psfMag',
```
whitespace after `16,`

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+                Values of faint magnitude limit.  Only points up to this limit will be plotted.
+                Beware of scrolling to too faint a limit; it might give you too many points!
+
+            filter_stream : explorer.plots.FilterStream, optional
```
-> ``qa.explorer.plots.FilterStream``? Should be in backticks. https://developer.lsst.io/python/numpydoc.html

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-        ----------
-        vdim : str
-            Name of dimension to explore.
+            ranges = {vdim : (y_min, y_max),
```
no space before `:`

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-        filter_stream : explorer.plots.FilterStream, optional
-            Stream of constraints that controls what data to display.  Useful to link
-            multiple plots together
+class MatchedQADataset(QADataset):
```
expected 2 blank lines

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
+    match_radius : `float`
+        Max match distance in arcsec.  Default is 0.5.
 
+    match_registry : `str`
```
What do you use this for in MatchedCatalog? OK if optional

timothydmorton replied:
This is a placeholder for potential caching of the results of matching, as I had done previously in `catalog.MatchedCatalog`.  I didn't port this idea into the new paradigm yet.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-                    x_range=None, y_range=None):
-        """Datashaded layout of color-color plots 
+    """
+    def __init__(self, data1, data2,
```
expected 1 blank line

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
 
-        xmax : float
-            faint magnitude limit.
+        ra1, dec1 = self.data1.df.ra[isPrimary1], self.data1.df.dec[isPrimary1]
```
Hmmm. I like that you're protecting people from being idiots, and this is the #1 user error with catalogs. But it seems restrictive. Can we talk about this?

timothydmorton replied:
Sure-- let me know what concerns you about it.

timothydmorton replied:
@yalsayyad just a ping to let me know if there's something you'd like me to address here.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
-            pts_list.append(dset.to(hv.Points, kdims=[c1, c2], groupby=[]).redim.range(**{c1:(-0.2,1.5),
-                                                                                          c2:(-0.2,1.5)}))
-        return hv.Layout([dynspread(datashade(pts, dynamic=False, x_range=x_range, y_range=y_range)) for pts in pts_list]).cols(2)
+        logging.info('{0} matched within {1} arcsec, {2} did not.'.format(good.sum(), self.match_radius, (~good).sum()))
```
line too long.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
```diff
@@ -1,806 +1,329 @@
 
-    def color_explore(self, xmax_range=np.arange(18,26.1,0.5), filter_stream=None):
-        """Dynamic map of color-color plots
+        match_df = pd.DataFrame({'id2':i2, 'distance':d}, index=i1)
```
`{'id2': i2, 'distance': d}`

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
What's this?

timothydmorton replied:
A placeholder for the fact that I suppose it's possible to imagine making a matched dataset and to look at some other quantity other than the difference between the two vdim values (e.g., mean).  Possibly won't ever get used like this, but it seemed reasonable to me.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
Maybe want the OR or the AND of the data1 and data2 flags?  I guess people could ask for that manually.

timothydmorton replied:
`flags` just contains a list of names of columns that are bools.  It is used to make sure that those dimensions are included in the `hv.Dataset`, which is subsequently used to implement the flag-select-y widgets.

yalsayyad commented on python/lsst/qa/explorer/dataset.py:
What is `dim[:-4]`?

timothydmorton replied:
It's the `"cmodel_mag"` part of `"cmodel_magDiff"`, e.g.  This is meant to deal with the rather specific situation of matching two `QADatasets` that have `*_magDiff` columns (e.g. `cmodel - psf` mags).  Rather than looking at the difference of the difference (which is a potentially confusing quantity), this changes to look at the difference between the datasets.   Again, it's a very specific situation and requires the `_magDiff` naming convention, but I did it because these `magDiff`-type columns are relatively common.  Probably the column naming convention for these magdiff functors should be documented somewhere.

yalsayyad commented on python/lsst/qa/explorer/display.py:
```diff
@@ -95,7 +96,7 @@ def getExp(self, ra, dec, **kwargs):
             ra, dec  = obj.ra, obj.dec
 
-        pos = afwCoord.IcrsCoord(ra*afwGeom.degrees, dec*afwGeom.degrees)
+        pos = afwGeom.SpherePoint(ra, dec, afwAngle.degrees)
```
You don't have to `import lsst.afw.geom.angle as afwAngle`. `afwGeom.degrees` works

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+"""Command-line task and associated config for writing QA tables.
```
https://developer.lsst.io/python/style.html#standard-code-order-should-be-followed
https://developer.lsst.io/stack/license-and-copyright.html#python-preamble

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+from lsst.coadd.utils import TractDataIdContainer
+from lsst.pipe.tasks.multiBand import MergeSourcesTask, MergeSourcesConfig
+from lsst.pipe.tasks.multiBand import _makeGetSchemaCatalogs
+from lsst.coadd.utils.coaddDataIdContainer import ExistingCoaddDataIdContainer
```
Almost all of these imports up to this point are not used.  Linters for text editors highlight these.

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+from lsst.utils import getPackageDir
+
+import functools
+import re, os
```
`re` not used

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+from .parquetTable import ParquetTable
+from .functors import CompositeFunctor, RAColumn, DecColumn, Column
+
+class PostprocessAnalysis(object):
```
expect 2 blank lines

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+
+        """
+        filt = dataId.get('filter', None)
+        funcs = CompositeFunctor.from_yaml(self.config.functorFile)
```
This disk access can be avoided and  I doubt the functorFile will be > 2GB.   Its should be read in `runDataRef`.

timothydmorton replied:
OK, so I'd read in `funcs` in `runDataRef` and pass it as an argument to `run`?

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+        self.write(df, patchRef)
+        return df
+
+    def run(self, parq, dataId):
```
`dataId` should be optional argument

timothydmorton replied:
But I require it in the run function, so how would that work?

yalsayyad replied:
We're trying to get all Gen2 stuff out of `run()` (@TallJimbo knows more)  Most run methods take a dataId so that the logger can print out the dataId info in the log... helps to know which `dataId` is failing!  So our compromise was to make `dataId` an optional keyword argument. Here's a suggestion for the inserting the patch ref.  It doesn't do the `filt` though. I guess I don't understand what the dataId's filter is for. It's typicallly 'merged' for your `deepCoadd_obj` table,  for `src` it'll be 'HSC-G' or 'HSC-R'. If the filter is specified in both the `parq` AND the yaml `funcs` definitions, what does the `filt` do in PostProcessAnalysis?

```
    def runDataRef(self, patchRef):
        parq = patchRef.get()
        translationDefinition = yaml.load(open(self.config.functorFile)
        df = self.run(parq, translationDefinition, dataId=patchRef.dataId)
        self.write(df, patchRef)
        return df

    def run(self, parq, translationDef, dataId=None):
        filt = dataId.get('filter', None)  #  Not sure wha to do with this!
        funcs = CompositeFunctor.from_yaml(translationDef)
        df = PostprocessAnalysis(parq, funcs, filt=filt).df
        if dataId is not None:
            for key, value in dataId.items():
                df[key] = value
        return df
```

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+        funcs = CompositeFunctor.from_yaml(self.config.functorFile)
+        analysis = PostprocessAnalysis(parq, funcs, filt=filt)
+        df = analysis.df
+        df['patchId'] = dataId['patch']
```
This seems too specific to your  `MultibandPostprocessTask` to be in the Base class's run method.  

 Maybe pass a series of additional identifier column labels.  For`MultibandPostprocessTask` it's {'patchId': 'patch'}.

In gerenal, we try to be dataId key agnostic throughout our tasks, hardcoding the input and output datatypes, but getting the keys that describe them from the butler.  We could do this programmatically here too.

timothydmorton replied:
Hmm, yes, good point.  I will probably ask for suggestions about how to do this one.

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+        df['patchId'] = dataId['patch']
+        return df
+
+
```
one blank line

yalsayyad commented on python/lsst/qa/explorer/postprocess.py:
```diff
@@ -0,0 +1,299 @@
+
+class WriteQATableConfig(PostprocessConfig):
+    def setDefaults(self):
+        self.functorFile = os.path.join(getPackageDir("qa_explorer"),
```
I'm not convinced that filepaths are the container of choice for the yaml

timothydmorton replied:
Should the default `QATable` specification YAML be a butler dataset?  Or defined in obs packages?  I'm open to better suggestions if we don't like it living next to the code like this.

yalsayyad replied:
I'm OK with the default living here. It's just for unit testing, right? The real ones will live their obs packages.

We generally don't specify filenames are in config parameters. Its OK to leave it temporarily.  I do want to be able to run this task without writing a file to disk and reading it back in though.

yalsayyad reviewed (commented):
Looked at dataset, display, and quickly at postprocess. Heading to sleep, but will pick up in the morning.  Delete display_old.

timothydmorton committed:
pep8 fixes

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -222,85 +332,279 @@ def columns(self):
+        return super().__call__(catalog, **kwargs)
 
 
 def fluxName(col):
```
This function doesn't work anymore. If I do
 ```
gPsFlux:
         functor: NanoMaggie
         args: slot_PsfFlux
         filt: HSC-G
         dataset: forced_src
``` 
I get "[('forced_src', 'HSC-G', 'slot_PsfFlux_flux')] not in index"  and if I do: 
``` 
gPsFlux:
        functor: NanoMaggie
        args: slot_PsfFlux_instFlux
        filt: HSC-G
        dataset: forced_src
``` 
I get 

"[('forced_src', 'HSC-G', 'slot_PsfFlux_instFlux_flux')] not in index" The column in the deepCoadd_obj table is actually called "'slot_PsfFlux_instFlux'" now to reflect that we are reporting instrumental fluxes.

timothydmorton replied:
aahh, yes.  I should be testing this function.

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -1,148 +1,255 @@
 
     @classmethod
-    def from_yaml(cls, filename):
+    def from_yaml(cls, filename, **kwargs):
```
We need a `from_yaml()` and a `from_file()`

timothydmorton replied:
meaning, `from_yaml()` to take a string as argument, and a `from_file()` that reads the file into a string and calls the `from_yaml()`?  Why do we need this?

yalsayyad replied:
So `run` doesn't have to read the file from disk.  I was imagining it'd work like:

```
with open(functorFilePath) as f:
    translationDefinition = yaml.safe_load(f)

CompositeFunction.from_yaml(translationDefinition)
```

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -160,11 +267,12 @@ def mag_aware_eval(df, expr):
         val = df.eval(expr_new, truediv=True)
-    except:
+    except Exception:  # Should check what actually gets raised
         expr_new = re.sub('mag\((\w+)\)', '-2.5*log(\g<1>_flux)/log(10)', expr)
```
Not sure this works anymore.

yalsayyad replied:
Do a search-replace for any use of `_flux` in your code. I think it'd be good to update your unit test file too to reflect our naming change.

timothydmorton replied:
yup, agreed.

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -222,85 +332,279 @@ def columns(self):
+
 class CoordColumn(Column):
-    """Base class for coordinate column
+    """Base class for coordinate column, in degress
```
degrees

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -222,85 +332,279 @@ def columns(self):
-        super(Mag, self).__init__(**kwargs)
+        self.calib = calib
+        if calib is not None:
+            self.fluxMag0 = calib.getFluxMag0()[0]
```
We need to assume there will be a fluxMag0 uncertainty

timothydmorton replied:
OK, but will that change the mag?  Does it need to be in the `Mag` functor? `fluxMag0Err` is handled in `MagErr`.

yalsayyad commented on python/lsst/qa/explorer/dataset_old.py:
```diff
@@ -0,0 +1,804 @@
+import numpy as np
```
I'm assuming this file wasn't meant to be in this PR. git rm.

yalsayyad commented on python/lsst/qa/explorer/functors.py:
```diff
@@ -1,148 +1,255 @@
+        try:
+            valDf = pd.concat(valDict, axis=1)
+        except TypeError:
+            print([(k, type(v)) for k, v in valDict.items()])
```
Should be an lsst logger.

yalsayyad commented on tests/testFunctors.py:
```diff
@@ -0,0 +1,184 @@
+import os
```
FIx standards compliance in whole file

yalsayyad replied:
And hurray for unit tests!

yalsayyad commented on tests/testFunctors.py:
```diff
@@ -0,0 +1,184 @@
+
+        func2 = Column('base_FootprintArea_value', filt='HSC-G')
+
+        ## The following fails with different dtypes:
```
remove commented out code

yalsayyad commented on tests/testFunctors.py:
```diff
@@ -0,0 +1,184 @@
+
+    def testMagDiff(self):
+        for filt in self.filters:
+            filt = 'HSC-G'
```
You're reassigning 'HSC-G'; not actually looping

timothydmorton committed:
add support for noDups and flattening multilevel result

This gives the option to have the output of a MultibandPostprocessingTask
to be a flat dataframe (as opposed to a df with multi-level column
index).  If this is the case, then the columns will be named with a
filter prefix (as defined by .filterMap), except for those functors
that are defined to be noDup=True.

made default noDups

make sure columns are not duplicated

fix flattening if filter is not there

timothydmorton committed:
write separate getFunctors and getAnalysis methods

This is useful for interactive notebook-based testing of tasks,
to inspect what the tasks would create for these objects.

timothydmorton committed:
define from_file in addition to from_yaml

This also takes disk access out of .run()

timothydmorton committed:
remove dataset_old.py

timothydmorton committed:
Add tests for MultibandPostprocessingTask

These tests ensure that the results are generated with the
proper column names, and also test the option to flatten/name-munge
the output tables to make sure that works properly.

timothydmorton committed:
_flux -> _instFlux

timothydmorton committed:
correct tests

timothydmorton committed:
remove debug

timothydmorton committed:
correct shortFilters creation in tests

timothydmorton committed:
calib_psfUsed -> calib_psf_used

timothydmorton committed:
fixup! calib_psfUsed -> calib_psf_used

timothydmorton committed:
make WriteQATableConfig inherit from MultibandPost...

timothydmorton committed:
update temp multilevel parq file

timothydmorton committed:
fix composite functor filt assignment?

timothydmorton committed:
remember to pass kwargs to from_yaml

timothydmorton committed:
make tests fail more verbosely on flattening results

timothydmorton committed:
add defaultFuncs such that ra, dec appear in .getFunctors() and thus in noDupCols

timothydmorton committed:
sort column index to check df compares

timothydmorton committed:
fixup! sort column index to check df compares

timothydmorton committed:
fixup! fixup! sort column index to check df compares

yalsayyad commented:
This ticket and PR was merged.
